### PR TITLE
Feature/40009

### PR DESCRIPTION
**Pull request by harshsuccessive**
Created an authmiddlerware.ts
to perform authentication and check authorization

### CODE REVIEW

File: extraTs/utils/validation.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { validateEmail } from './helpers';
+import {IUsers} from '../interfaces'
   export default function validateUser ( user: IUsers[] ): void {
```
Comment (namanpar) [inline extraTs/utils/validation.ts:3]: use es6

File: src/controllers/trainee/controller.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+             });
+
+        }catch(err){
+            console.log("inside err", err);
+
+        }
+
+    }
+
+    Delete( req: Request, res: Response, next: NextFunction ) {
```
Comment (namanpar) [inline src/controllers/trainee/controller.ts:75]: proper function naming

File: src/libs/permission.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                    else {
+                        continue;
+                    }
+
+                }
+            }
+        }
+        else {
+            continue;
+        }
```
Comment (namanpar) [inline src/libs/permission.ts:26]: improve this function logic and code, use es6

File: src/libs/routes/authMiddleWare.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    const token = req.headers['authorization'];
+    const decodedUser =  jwt.verify(token, 'qwertyuiopasdfghjklzxcvbnm123456');
+    const { role } = decodedUser
+    console.log( 'hello ' );
+    console.log( 'User', decodedUser );
+    
+    if ( role ) {
+        if ( hasPermission( module, role, permissionType )) {
+            next();
+        }
```
Comment (namanpar) [inline src/libs/routes/authMiddleWare.ts:19]: name the given function

**Review by namanpar — CHANGES_REQUESTED** [review]
resolve conflict and do the requested changes

File: extraTs/utils/validation.ts (hunk 1 above)
Comment (harshsuccessive) [reply extraTs/utils/validation.ts:3]: Shifted to es6

File: src/controllers/trainee/controller.ts (hunk 2 above)
Comment (harshsuccessive) [reply src/controllers/trainee/controller.ts]: changed the names

File: src/libs/permission.ts (hunk 3 above)
Comment (harshsuccessive) [reply src/libs/permission.ts]: changed the logic and code and shifted to es6

File: src/libs/routes/authMiddleWare.ts (hunk 4 above)
Comment (harshsuccessive) [reply src/libs/routes/authMiddleWare.ts]: named it

File: src/libs/permission.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { permissions }  from './constant';
+
+export default function hasPermission(moduleName: string, role: string, permissionType: string): boolean {
+    const permission = permissions[moduleName];
+    if (permission.all.includes(role)) {
```
Comment (namanpar) [inline src/libs/permission.ts:5]: declare method in es6

**Review by namanpar — CHANGES_REQUESTED** [review]
Fix lint and do requested changes

File: src/libs/permission.ts (hunk 5 above)
Comment (harshsuccessive) [reply src/libs/permission.ts]: done

**Review by namanpar — APPROVED** [review]

**Pull request merged**